docs/advanced/checkpointing.md

```diff
@@ -0,0 +1,108 @@
+
+## Example Failure Scenarios
+Below are some examples of what can go wrong during processing and how application will be recovering from it.  
+In all the cases, the application stops, and it needs to be restarted.
+
+- **Error happens during processing of the message**.  
+The app will stop without producing changelogs and committing offsets.  
+On restart, it will restart processing from the latest committed offset.
+
+- **Checkpointing fails to produce some state updates to the changelog topics.**   
+Input topic offsets will not be committed, and the local state stores will not be flushed.  
+On recovery, application will ignore the changelog updates belonging to non-committed input topic offsets.  
+After recovery, the app will restart processing from the latest committed offset.
+
+- **Checkpointing fails to commit offsets to the input topics.**   
+Input topic offsets will not be committed, and the local state stores will not be flushed.  
+On recovery, application will ignore the changelog updates belonging to non-committed input topic offsets.  
+After recovery, the app will restart processing from the latest committed offset.  
+
+- **Checkpointing fails to flush state stores to the disk.**  
+All changelog updates are produced, and the input topic offsets are committed. 
+During recovery, the app will apply changelog updates to the local state stores.
+
+## Configuring the Checkpointing
+
+Users may configure how often the checkpoints are committed by passing the `commit_interval` parameter to the `Application` class.
+
+The default commit interval is 5 seconds.
+
+Changing the commit interval will have several implications for the application:
+
+- ***Longer commit intervals*** will make the application flush state stores and changelogs less often, limiting the IO.
+    
+    At the same time, it may lead to larger memory usage because more state updates will be accumulated in memory.
+    
+    The required amount of memory depends on how many unique keys are processed (state transactions batch updates per key).
+    
+    Also, when an application fails, it will need to reprocess more messages since the latest checkpoint, increasing the number of duplicates produced to the output topics.
+    
+- ***Shorter commit intervals*** will lead to more IO because changelogs and state stores will be flushed more frequently.
+    
+    However, it will reduce memory usage and limit the number of potentially reprocessed messages, reducing duplicates.
+    
+- If `commit_interval` is set to `0`, the application will commit a checkpoint for every processed Kafka message.
+
+When configuring the commit interval, take into account such factors as the number of unique keys in the input topics, hardware, and infrastructure.
+
+## Limitations
+
+In the At-Least-Once setting, it is still possible that unwanted changelog changes get applied during recovery from scratch. 
```

Comment:
Might be worth mentioning what the user can do to help avoid this situation, even if it might be obvious? 

ex: "don't update code without confirming the consumer successfully committed its checkpoint".

Reply:
Good call, I added a recommendation.

The best way to avoid it is, ofc, EOS